// @rpldy/upload-preview/lib/cjs/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getWithMandatoryOptions = exports.getFileObjectUrlByType = exports.getFallbackUrlData = void 0;
Object.defineProperty(exports, "isFunction", {
  enumerable: true,
  get: function () {
    return _shared.isFunction;
  }
});
var _shared = require("@rpldy/shared");
var _consts = require("./consts");
var _defaults = require("./defaults");
const getWithMandatoryOptions = options => {
  return {
    ..._defaults.PREVIEW_DEFAULTS,
    ...options
  };
};
exports.getWithMandatoryOptions = getWithMandatoryOptions;
const getFallbackUrlData = (fallbackProp, file) => {
  let data = (0, _shared.isFunction)(fallbackProp) ? fallbackProp(file) : fallbackProp;
  if (typeof data === "string") {
    data = {
      url: data,
      name: file.name,
      type: _consts.PREVIEW_TYPES.IMAGE
    };
  }
  return data;
};
exports.getFallbackUrlData = getFallbackUrlData;
const getFileObjectUrlByType = (type, mimeTypes, max, file) => {
  let data;
  if (mimeTypes && ~mimeTypes.indexOf(file.type)) {
    if (!max || file.size <= max) {
      data = {
        url: URL.createObjectURL(file),
        name: file.name,
        type
      };
    }
  }
  return data;
};
exports.getFileObjectUrlByType = getFileObjectUrlByType;
